docs/item.rst

```diff
@@ -0,0 +1,200 @@
+
+``--stockpiled``
+    Only include items that are in stockpiles. Does not include empty bins,
+    barrels, and wheelbarrows assigned as storage and transport for stockpiles.
+
+``--scattered``
+    Opposite of ``--stockpiled``
+
+``--forbidden``
+    Only include forbidden items.
+
+``--melting``
+    Only include items designated for melting.
+
+``--dumping``
+    Only include items designated for dumping.
+
+``--visible``
+    Only include visible items (i.e., ignore hidden items).
+
+``--hidden``
+    Only include hidden items.
+
+
+API
+---
+
+The item script can be called programmatically by other scripts, either via the
+commandline interface with ``dfhack.run_script()`` or via the API functions
+defined in :source-scripts:`item.lua`, available from the return value of
+``reqscript('item')``:
+
+* ``item.execute(action, conditions, options)``
+* ``item.executeWithPrinting(action, conditions, options)``
+
+Performs ``action`` (``forbid``, ``melt``, etc.) on all items satisfying
+``conditions`` (a table containing functions from item to boolean). ``options``
+is a table containing the boolean flags ``artifact``, ``dryrun``, and
+``bytype``, which correspond to the (filter) options described above.
+
+The function ``item.execute`` performs no output, while the ``WithPrinting``
+variant performs the same output as the ``item`` tool.
+
+The API provides a number of helper functions to aid in the construction of the
+filter table. The first argument ``tab`` is always the table to which the filter
+should be added.
+
+* ``item.condition_burrow(tab,burrow, outside)``
+    Corresponds to ``--inside`` or ``--outside`` (when ``outside=true``). The
+    ``burrow`` argument must be a burrow object, not a string.
+
+* ``item.condition_type(tab, match)``
+    If ``match`` is a string, this corresponds to ``--type <match>``. Also
+    accepts numbers, matching against ``item:getType()``
+
+* ``item.condition_reachable(tab)``
+    Corresponds to ``--reachable``
+
+* ``item.condition_unreachable(tab)``
+    Corresponds to ``--unreachable``
+
+* ``item.condition_description(tab, pattern)``
+    Corresponds to ``--description <pattern>``
+
+* ``item.condition_material(tab, match)``
+    Corresponds to ``--material <match>``
+
+* ``item.condition_matcat(tab, match)``
+    Corresponds to ``--mat-category <match>``
+
+* ``item.condition_wear(tab, lower, upper)``
+    Selects items with wear level between ``lower`` and ``upper`` (Range 0-3, see above).
+
+* ``item.condition_quality(tab, lower, upper)``
+    Selects items with quality between ``lower`` and ``upper`` (Range 0-5, see above).
+
+* ``item.condition_stockpiled(tab, invert)``
+    Selects stockpiled items, or scattered items when ``invert=true``.
+
+* ``item.condition_forbidden(tab)``
+    Checks for ``item.flags.forbid``
+
+* ``item.condition_melt(tab)``
+    Checks for ``item.flags.melt``
+
+* ``item.condition_dump(tab)``
+    Checks for ``item.flags.dump``
```

Comment:
shouldn't these take a boolean so the inverse can be matched?

Reply:
also, there seem to be a mix of `condition.foo()` + `condition.notfoo()` and `condition.foo(bool)`. I'd prefer to standardize on the latter, taking a boolean to indicate whether the relevant property should be true or false. that is foo(true) should match foo being true. I'd prefer to avoid invert=true semantics, like what is currently on the `stockpiled` condition.

Reply:
I agree, the API for condition creation still looks somewhat haphazard and that `condition.foo(...,bool)` is the way to go. However, I would still advocate for `negate=true` semantics. This way, the argument can be naturally elided when the condition is to be matched positively, which is probably the majority of cases. Otherwise, this would require `true` and `nil` to behave the same and `false` to stand for negation.